### DISCUSSION TITLE
how to link libc liberty

### DISCUSSION THREAD

**Discussion by changpenghui616** [memtest86plus/memtest86plus#100] [Q&A] [answered]
I need to add a function to main.c for test，use the ‘fopen’ ，and already add #include <stdio.h> in main.c， when build project  display “main.c:(.text+0xcc): undefined reference to `fopen'”
i try to add ‘-lc’ to maikefile，like this
![image](https://user-images.githubusercontent.com/106718857/172806231-43c6c4f0-8dd4-490b-ae23-570b9440a4ef.png)

but it doesn't work

**Comment by martinwhitaker** [comment, marked as answer]
Memtest86+ is a bare metal program. You can't use a library that requires an underlying operating system.

**Reply by changpenghui616** [reply]
thanks reply
I try to enable PC enter windows system automatic after memtest86 runs successfully， so i try to when memtes86 running， change the boot/.efi to the original file（use libc function）， whether have other way if cant use this library？
PS：whether can use UEFI protocol（EFI_SIMPLE_FILE_SYSTEM_PROTOCOL）in “efisetup.c” achieve this target？

**Reply by x86fr** [reply]
Memtest86+ can't access any FAT volume at this time. 
It will probably be added at some time, but not anytime soon as this require a lot of code. 

What you try to achieve may be doable by using UEFI scripting or a specific PXE script. Just add a function to Memtest86+ to automatically reboot after a successful pass and use UEFI scripting to put a flag somewhere if Memtest86+ has already been run previously.

**Reply by changpenghui616** [reply]
thanks，
 i find Memtest86+ rewrite some  interface about UEFI， so it support UEFI boot， but no dependence any liberty.  do I understand this correctly？
so i create a efi application（call it FileOperate.efi） from EDK2 framework（can use file interface from EDK2），it can access current work directory， i guess it can realize File operation if  as a bootx64.efi when boot。
As you say， if try to use UEFI script， how can i combine FileOperate.efi  and memtest.efi.  in other words, when PC boot, can run memtest.efi and FileOperate.efi that according to the order。
shall i create UEFI shell application as boot.efi， then call memtest.efi？

**Comment by changpenghui616** [comment]
use uefi shell can achive this ，thans for help @martinwhitaker @x86fr
